silq/analysis/analysis.py

```diff
@@ -715,6 +717,38 @@ def analyse_EPR(empty_traces: np.ndarray,
             'mean_high_blip_duration': results_read['mean_high_blip_duration']}
 
 
+def analyse_threshold_up_proportion(up_proportions_arrs: np.ndarray,
+                                    shots_per_frequency: int):
+    proportion_space = np.linspace(0, 1, num=shots_per_frequency + 1,
+                                   endpoint=True)
+    up_proportions_arrs = up_proportions_arrs.reshape(1, -1)
+    kernel = gaussian_kde(up_proportions_arrs)
+    gaussian_up_proportions = kernel(proportion_space)
+    peak_idxs = peakutils.peak.indexes(gaussian_up_proportions,
+                                       thres=0.5/up_proportions_arrs.shape[-1],
+                                       min_dist=shots_per_frequency/5)
+    if len(peak_idxs) == 0:
+        logger.debug(f'Adaptive thresholding routine: 0 peaks were '
+                     f'found, using threshold 0.5')
+        true_idx = round((len(proportion_space) - 1) / 2)
+    elif len(peak_idxs) == 1:
+        if (len(proportion_space) - peak_idxs[0]) / len(proportion_space) > 0.5:
+            true_slice = slice(*[peak_idxs[0], len(proportion_space)])
+            true_idx = np.argmin(np.round(gaussian_up_proportions[true_slice], 4)) + \
+                       peak_idxs[0]
+        else:
+            true_slice = slice(*[0, peak_idxs[0]])
+            true_idx = np.argmin(np.round(gaussian_up_proportions[true_slice], 4))
+    else:
+        true_slice = slice(*[min(peak_idxs), max(peak_idxs)])
+        true_idx = np.argmin(np.round(gaussian_up_proportions[true_slice], 4)) + \
```

Comment:
Could you explain the rounding please?
Is it necessary?

Reply:
Since the minimum value of gaussian distribution sometimes gets to 1e-34 at spin-up proportion of 1 or 0 - without rounding we will get threshold = 1 or 0. But the idea is to potentially use the threshold as a function of charge stability (any jumps in spin-up proportion can be detected) throughout our long measurements. Thus, when we don't have two gaussian peaks (but just one when nucleus is not flipping), we will have weird 0 or 1 threshold. To eliminate these jumps by rounding to the 4 digit after point, will make threshold 0.6 instead of 1 for example. It is not elegant solution, but I haven't found yet a better one.